pyomo/core/base/var.py

```diff
@@ -371,70 +354,169 @@ def __getstate__(self):
     # Abstract Interface
     #
 
-    # value is an attribute
+    def set_value(self, val, valid=False):
+        """Set the current variable value.
+
+        Set the value of this variable.  The incoming value is converted
+        to a numeric value (i.e., expressions are evaluated).  If the
+        variable has units, the incoming value is converted to the
+        correct units before storing the value.  The final value is
+        checked against both the variable domain and bounds, and an
+        exception is raised if the value is not valid.  Domain and
+        bounds checking can be bypassed by setting the ``valid``
+        argument to `True`.
+
+        """
+        # Special case: setting a variable to None "clears" the variable.
+        if val is None:
+            self._value = None
+            self.stale = True
+            return
+        # TODO: generate a warning/error:
+        #
+        # Check if this Var has units: assigning dimensionless
+        # values to a variable with units should be an error
+        if type(val) not in native_numeric_types:
+            if self.parent_component()._units is not None:
+                _src_magnitude = value(val)
+                _src_units = units.get_units(val)
+                val = units.convert_value(
+                    num_value=_src_magnitude, from_units=_src_units,
+                    to_units=self.parent_component()._units)
+            else:
+                val = value(val)
+
+        if not valid:
+            if val not in self.domain:
+                # logger.warning(
```

Comment:
Do we need this commented out section?

Reply:
For now, yes: this is holding the (hopeful) implementation of #2180.